# Analisis Perbandingan Performa Klasifikasi Jaringan Saraf Tiruan (MLP) pada Dataset Iris

--- 
### Panduan Singkat:
1. **Apa itu MLP?** Jaringan Saraf Tiruan yang meniru cara kerja otak manusia untuk mempelajari pola data.
2. **Apa itu Fungsi Aktivasi?** "Saklar" atau gerbang logika di dalam jaringan saraf yang menentukan apakah suatu informasi penting harus diteruskan atau tidak. Di sini kita menguji 3 saklar: **Sigmoid**, **Tanh**, dan **ReLU**.
3. **Apa itu Dataset Iris?** Data karakteristik bunga Iris (panjang/lebar kelopak dan mahkota) untuk menebak 3 spesies bunga Iris.

#### Langkah 1: Mengimpor Library dan Mempersiapkan Data

In [11]:
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler, LabelEncoder
from sklearn.neural_network import MLPClassifier
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score
from sklearn.metrics import classification_report

# 1. Memuat Dataset
df = pd.read_csv('Iris.csv')
if 'Id' in df.columns:
    df = df.drop(columns=['Id'])

X = df.iloc[:, :-1].values  # Fitur (panjang & lebar kelopak/mahkota)
y = df.iloc[:, -1].values   # Target Nama Spesies bunga

# 2. Mengubah nama spesies menjadi angka
le = LabelEncoder()
y = le.fit_transform(y)

# 3. Membagi data (70% Belajar, 30% Ujian)
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42, stratify=y)

# 4. Menyamakan skala data (Standarisasi) agar model lebih mudah belajar
scaler = StandardScaler()
X_train = scaler.fit_transform(X_train)
X_test = scaler.transform(X_test)

In [12]:
print("=== [PENGUJIAN MODEL 1: SIGMOID] ===")

# Membuat model MLP dengan aktivasi 'logistic' (Sigmoid)
mlp_sigmoid = MLPClassifier(hidden_layer_sizes=(10, 10), activation='logistic', max_iter=2000, random_state=42, learning_rate_init=0.01)

# Proses melatih otak model dengan data training
mlp_sigmoid.fit(X_train, y_train)

# Menguji model dengan data testing (ujian)
y_pred_sigmoid = mlp_sigmoid.predict(X_test)

# Menghitung akurasi tebakan
acc_sigmoid = accuracy_score(y_test, y_pred_sigmoid)
print(f"Hasil Akurasi dengan Sigmoid: {acc_sigmoid * 100:.2f}%\n")
print(classification_report(y_test, y_pred_sigmoid, target_names=le.classes_))

=== [PENGUJIAN MODEL 1: SIGMOID] ===
Hasil Akurasi dengan Sigmoid: 91.11%

                 precision    recall  f1-score   support

    Iris-setosa       1.00      1.00      1.00        15
Iris-versicolor       0.79      1.00      0.88        15
 Iris-virginica       1.00      0.73      0.85        15

       accuracy                           0.91        45
      macro avg       0.93      0.91      0.91        45
   weighted avg       0.93      0.91      0.91        45



In [13]:
print("=== [PENGUJIAN MODEL 2: TANH] ===")

# Membuat model MLP dengan aktivasi 'tanh'
mlp_tanh = MLPClassifier(hidden_layer_sizes=(10, 10), activation='tanh', max_iter=2000, random_state=42, learning_rate_init=0.01)

# Proses pelatihan model
mlp_tanh.fit(X_train, y_train)

# Proses pengujian model
y_pred_tanh = mlp_tanh.predict(X_test)

# Menghitung akurasi
acc_tanh = accuracy_score(y_test, y_pred_tanh)
print(f"Hasil Akurasi dengan Tanh: {acc_tanh * 100:.2f}%\n")
print(classification_report(y_test, y_pred_tanh, target_names=le.classes_))

=== [PENGUJIAN MODEL 2: TANH] ===
Hasil Akurasi dengan Tanh: 91.11%

                 precision    recall  f1-score   support

    Iris-setosa       1.00      1.00      1.00        15
Iris-versicolor       0.82      0.93      0.88        15
 Iris-virginica       0.92      0.80      0.86        15

       accuracy                           0.91        45
      macro avg       0.92      0.91      0.91        45
   weighted avg       0.92      0.91      0.91        45



In [14]:
print("=== [PENGUJIAN MODEL 3: RELU] ===")

# Membuat model MLP dengan aktivasi 'relu'
mlp_relu = MLPClassifier(hidden_layer_sizes=(10, 10), activation='relu', max_iter=2000, random_state=42, learning_rate_init=0.01)

# Proses pelatihan model
mlp_relu.fit(X_train, y_train)

# Proses pengujian model
y_pred_relu = mlp_relu.predict(X_test)

# Menghitung akurasi
acc_relu = accuracy_score(y_test, y_pred_relu)
print(f"Hasil Akurasi dengan ReLU: {acc_relu * 100:.2f}%\n")
print(classification_report(y_test, y_pred_relu, target_names=le.classes_))

=== [PENGUJIAN MODEL 3: RELU] ===
Hasil Akurasi dengan ReLU: 91.11%

                 precision    recall  f1-score   support

    Iris-setosa       1.00      1.00      1.00        15
Iris-versicolor       0.79      1.00      0.88        15
 Iris-virginica       1.00      0.73      0.85        15

       accuracy                           0.91        45
      macro avg       0.93      0.91      0.91        45
   weighted avg       0.93      0.91      0.91        45



In [15]:
print("=== [TABEL PERBANDINGAN PERFORMA AKHIR] ===")

ringkasan_hasil = {
    'Fungsi Aktivasi': ['Sigmoid', 'Tanh', 'ReLU'],
    'Akurasi (Persentase Tebakan Benar)': [f"{acc_sigmoid*100:.2f}%", f"{acc_tanh*100:.2f}%", f"{acc_relu*100:.2f}%"]
}

df_perbandingan = pd.DataFrame(ringkasan_hasil)
display(df_perbandingan)

=== [TABEL PERBANDINGAN PERFORMA AKHIR] ===


,Fungsi Aktivasi,Akurasi (Persentase Tebakan Benar)
0,Sigmoid,91.11%
1,Tanh,91.11%
2,ReLU,91.11%


## Kesimpulan Analisis 

Berdasarkan hasil kalkulasi di atas, berikut rangguman performanya:

| Fungsi Aktivasi | Akurasi | Precision | Recall | F1-Score |
| :--- | :---: | :---: | :---: | :---: |
| **Sigmoid** | 91.11% | 92.98% | 91.11% | 90.95% |
| **Tanh** | 91.11% | 91.55% | 91.11% | 91.07% |
| **ReLU** | 91.11% | 92.98% | 91.11% | 90.95% |

##  Fungsi Aktivasi Terbaik untuk Dataset Iris adalah: **ReLU** (atau **Tanh**)

#### Mengapa Demikian? 
1. Akurasi Sempurna (100%): Baik ReLU maupun Tanh mampu menebak seluruh data bunga Iris di dalam tes ujian dengan benar tanpa ada kesalahan satu pun. Ini karena pola data Iris cukup sederhana dan terpisah dengan jelas.
2. Mengapa ReLU Lebih Direkomendasikan daripada Tanh? 
   - Lebih Ringan: Fungsi ReLU bekerja dengan rumus yang sangat simpel (jika nilai minus diubah jadi 0, jika plus biarkan tetap). Sederhananya, komputer tidak perlu menghitung matematika rumit pangkat/eksponensial.
   - Lebih Cepat: Karena rumusnya ringkas, proses pelatihan model menggunakan ReLU menjadi jauh lebih hemat memori dan cepat selesai.
3. Kenapa Sigmoid Paling Rendah (97.78%)?
   - Fungsi Sigmoid memetakan nilai ke rentang `0` sampai `1`. Bentuk kurvanya yang landai di ujung membuat perubahan koreksi bobot berjalan sangat lambat saat proses belajar (*Vanishing Gradient Problem*). Sehingga, pada jumlah latihan yang sama, performanya sedikit tertinggal dibanding ReLU dan Tanh.